pfsimulator/parflow_lib/parameter_union.h

```diff
@@ -0,0 +1,233 @@
+/*BHEADER**********************************************************************
+*
+*  Copyright (c) 1995-2024, Lawrence Livermore National Security,
+*  LLC. Produced at the Lawrence Livermore National Laboratory. Written
+*  by the Parflow Team (see the CONTRIBUTORS file)
+*  <parflow@lists.llnl.gov> CODE-OCEC-08-103. All rights reserved.
+*
+*  This file is part of Parflow. For details, see
+*  http://www.llnl.gov/casc/parflow
+*
+*  Please read the COPYRIGHT file or Our Notice and the LICENSE file
+*  for the GNU Lesser General Public License.
+*
+*  This program is free software; you can redistribute it and/or modify
+*  it under the terms of the GNU General Public License (as published
+*  by the Free Software Foundation) version 2.1 dated February 1999.
+*
+*  This program is distributed in the hope that it will be useful, but
+*  WITHOUT ANY WARRANTY; without even the IMPLIED WARRANTY OF
+*  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the terms
+*  and conditions of the GNU General Public License for more details.
+*
+*  You should have received a copy of the GNU Lesser General Public
+*  License along with this program; if not, write to the Free Software
+*  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307
+*  USA
+**********************************************************************EHEADER*/
+
+/*
+ * ParameterUnion looks for a parameter that might go by
+ * different aliases or assume different types of values.
+ *
+ * It can be used in cases where:
+ * - a physical variable may come under different names
```

Comment:
What is the use case for aliasing?   Where are we doing this?

This will make input correctness much more complicated; does the python error checking support this?

Reply:
Simplifying the coding for the Constant/PFBFile etc cases is good.

Part of the reason for explicitly setting the type explicit is this allows using the same attribute names for different types.   If new classes are added that also have same attributes then the attribute names alone wouldn't be sufficient unless they are named differently.   E.G. in the future if another type is added that also takes Value or FileName attributes.    Using simple attribute names is no longer future proof and I think this would require the attribute names be Type specific to disambiguate.

I'm not fully understanding how the union simplifies case of DeepAquiferBC being set from permeability, wouldn't you need a key to single that, like a Type key?   Key.Type = Permeability? 

I need to look over the python error checking, they wrote a little domain specific language to define the keys and I'm not sure it would handle aliasing.   I think they might use the Type attribute as well to check.

Reply:
The Python error checking does not appear to be looking at the type and then checking to make sure that required attributes for that type are present.   It probably should do that.

Reply:
Just thinking out loud....another item discussed at the start of the Python was enabling auto-complete in an editor for helping to build input files.   I wonder if having a Type would make that process flow better.   User auto-completes the Type and then auto-complete could know which attributes are required.

Reply:
I see. I understand the reasoning now. 

As for the permeability, my reasoning was that Parflow could be flexible enough not to expect any specific combination of input parameters, and display different behaviors depending on the information that it was provided. Permeability would be an additional optional key to provide futher customization of inputs. My idea was to generalize the union even further to allow a parameter defined from a combination of others. I am starting to realize that while this reduces the size of the Tcl and python scripts, it might not necessarily improve user experience.

What do you think is the best approach moving forward, having into consideration the future intended developements for the `DeepAquiferBC`?

Reply:
Can you elaborate on your comment about the Python error checking? Is this something to do with the `YAML` generated files for the python format?